File: nim/05-functions/recursion.nim
# Factorial: n! = n * (n-1)!
proc factorial(n: int): int =
  if n <= 1:
    1
  else:
    n * factorial(n - 1)

# Fibonacci: each number is the sum of the previous two
proc fib(n: int): int =
  if n < 2:
    n
  else:
    fib(n - 1) + fib(n - 2)

echo "factorial(5) = ", factorial(5)
echo "fib(10) = ", fib(10)

File: php/05-functions/recursion.php
<?php
declare(strict_types=1);

// factorial(n) = n * factorial(n - 1), with factorial(0) = 1
function factorial(int $n): int
{
    if ($n <= 1) {
        return 1; // base case
    }
    return $n * factorial($n - 1); // recursive case
}

// The Fibonacci sequence: each number is the sum of the previous two
function fibonacci(int $n): int
{
    if ($n < 2) {
        return $n; // base cases: fib(0) = 0, fib(1) = 1
    }
    return fibonacci($n - 1) + fibonacci($n - 2);
}

echo "5! = " . factorial(5) . "\n";
echo "10! = " . factorial(10) . "\n";

echo "First 10 Fibonacci numbers: ";
for ($i = 0; $i < 10; $i++) {
    echo fibonacci($i) . " ";
}
echo "\n";
